glib/src/boxed_any_object.rs

```diff
@@ -0,0 +1,133 @@
+    #[error("Can't convert item inside BoxedAnyObject to requested type")]
+    InvalidType,
+    #[error("Item already immutably borrowed")]
+    AlreadyBorrowed(#[from] std::cell::BorrowError),
+    #[error("Item already mutably borrowed")]
+    AlreadyMutBorrowed(#[from] std::cell::BorrowMutError),
+}
+
+mod imp {
+    use super::*;
+
+    #[derive(Debug)]
+    pub struct BoxedAnyObject {
+        pub item: RefCell<Box<dyn Any>>,
+    }
+    #[object_subclass]
+    impl ObjectSubclass for BoxedAnyObject {
+        const NAME: &'static str = "BoxedAnyObject";
+        type Type = super::BoxedAnyObject;
+    }
+    impl Default for BoxedAnyObject {
+        fn default() -> Self {
+            Self {
+                item: RefCell::new(Box::new(None::<usize>)),
+            }
+        }
+    }
+    impl ObjectImpl for BoxedAnyObject {}
+}
+
+wrapper! {
+    /// # Examples
+    /// ```rust
+    /// use glib::BoxedAnyObject;
+    /// struct Author {
+    ///     name: String,
+    ///     subscribers: usize
+    /// }
+    /// // BoxedAnyObject can contain any custom type
+    /// let boxed = BoxedAnyObject::new(Author {
+    ///     name: String::from("GLibAuthor"),
+    ///     subscribers: 1000
+    /// });
+    ///
+    /// // The boxed data can be stored as a GObject
+    /// let list = gio::ListStore::new(BoxedAnyObject::static_type());
+    /// list.append(boxed.clone().upcast());
+    ///
+    /// // And can be retrieved with `borrow`
+    /// let author: Author = boxed.borrow();
+    /// ```
+    pub struct BoxedAnyObject(ObjectSubclass<imp::BoxedAnyObject>);
+}
+
+impl Default for BoxedAnyObject {
+    fn default() -> Self {
+        Object::new(&[]).expect("Failed to create BoxedAnyObject")
+    }
+}
+impl BoxedAnyObject {
+    pub fn new<T: 'static>(item: T) -> Self {
+        let obj = Self::default();
+        obj.replace(item);
+        obj
+    }
+
+    /// Replaces the wrapped value with a new one, returning the old value, without deinitializing either one.
+    pub fn replace<T: 'static>(&self, t: T) -> Box<dyn Any> {
+        self.impl_().item.replace(Box::new(t) as Box<dyn Any>)
+    }
+
+    /// Immutably borrows the wrapped value, returning none if the value is currently mutably borrowed or
+    /// the inner value has never been seet
+    pub fn try_borrow<T: 'static>(&self) -> Result<Ref<'_, T>, BorrowError> {
+        // The required function is only available on nightly:
+        // https://doc.rust-lang.org/std/cell/struct.Ref.html#method.filter_map.
+        // As a workaround, I check if everything is safe, then I unwrap
+
+        let borrowed = self.impl_().item.try_borrow()?;
+        borrowed
+            .as_ref()
+            .downcast_ref::<T>()
+            .ok_or(BorrowError::InvalidType)?;
+        Ok(self.borrow()) // Now this won't panic
+    }
+
+    /// Mutably borrows the wrapped value, returning none if the value is currently immutably borrowed or
+    /// the inner value has never been seet
```

Comment:
As above

Reply:
And also below

Reply:
And the dot at the end too.

Reply:
Ups, that line shouldn't even be there, since `Option` is not a problem anymore. There's always some value inside...
I have to remember Rust checks the code, but not the comments 😆